static/src/javascripts/projects/commercial/modules/messenger.ts

```diff
@@ -0,0 +1,382 @@
+		'id' in payloadToCheck &&
+		payloadToCheck.type in LISTENERS &&
+		/^[a-f0-9]{8}-([a-f0-9]{4}-){3}[a-f0-9]{12}$/.test(payloadToCheck.id)
+	);
+};
+
+/**
+ * Cheap string formatting function
+ *
+ * @param error An object `{ code, message }`. `message` is a string where successive
+ * occurrences of %% will be replaced by the following arguments
+ * @param args Arguments that will replace %%
+ *
+ * @example
+ * formatError({ message: "%%, you are so %%" }, "Regis", "lovely")
+ * => { message: "Regis, you are so lovely" }
+ */
+const formatError = (error: { message: string }, ...args: string[]) =>
+	args.reduce((e, arg) => {
+		e.message = e.message.replace('%%', arg);
+		return e;
+	}, error);
+
+/**
+ * Convert a posted message to our StandardMessage format
+ *
+ * @param event The message event received on the window
+ * @returns A message with the `StandardMessage` format, or null if the conversion was unsuccessful
+ */
+const eventToStandardMessage = (event: MessageEvent) => {
+	try {
+		// Currently all non-string messages are discarded here since parsing throws an error
+		// TODO Review whether this is the desired outcome
+		const data: unknown = JSON.parse(event.data);
+		return isProgrammaticMessage(data)
+			? toStandardMessage(data)
+			: isValidPayload(data)
+			? data
+			: null;
+	} catch (ex) {
+		return null;
+	}
+};
+
+/**
+ * Callback that is fired when an arbitrary message is received on the window
+ *
+ * @param event The message event received on the window
+ */
+const onMessage = (
+	event: MessageEvent<string>,
+): Promise<unknown> | undefined => {
+	const data = eventToStandardMessage(event);
+
+	if (data === null) {
+		return;
+	}
+
+	// Respond to the original iframe with the result of calling the persistent listener / listener chain
+	const respond = (error: { message: string } | null, result: unknown) => {
+		postMessage(
+			{
+				id: data.id,
+				error,
+				result,
+			},
+			event.source ?? window,
+		);
+	};
+
+	const listener = LISTENERS[data.type];
+
+	if (Array.isArray(listener) && listener.length) {
+		// Because any listener can have side-effects (by unregistering itself),
+		// we run the promise chain on a copy of the `LISTENERS` array.
+		// Hat tip @piuccio
+		const promise =
+			// We offer, but don't impose, the possibility that a listener returns
+			// a value that must be sent back to the calling frame. To do this,
+			// we pass the cumulated returned value as a second argument to each
+			// listener. Notice we don't try some clever way to compose the result
+			// value ourselves, this would only make the solution more complex.
+			// That means a listener can ignore the cumulated return value and
+			// return something else entirely—life is unfair.
+			// We don't know what each callack will be made of, we don't want to.
+			// And so we wrap each call in a promise chain, in case one drops the
+			// occasional fastdom bomb in the middle.
+			listener.slice().reduce<Promise<unknown>>(
```

Comment:
Is this slice helpful here? `reduce` does not mutate the object it is called on.

> `reduce` does not directly mutate the object on which it is called but the object may be mutated by the calls to callbackfn. [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce)